pulsar-client-go/pulsar/client.go

```diff
@@ -101,17 +101,17 @@ type ClientOptions struct {
 type Client interface {
 	// Create the producer instance
 	// This method will block until the producer is created successfully
-	CreateProducer(ProducerOptions) (Producer, error)
+	CreateProducer(ProducerOptions, Schema) (Producer, error)
```

Comment:
For backward compatibility, can we do:

```
CreateProducer(ProducerOptions) (Producer, error)

CreateProducer(ProducerOptions, Schema) (Producer, error)
```

If we can't do that, I would do followings:

```
CreateProducer(ProducerOptions) (Producer, error)

CreateTypedProducer(ProducerOptions, Schema) (Producer, error)
```

Reply:
I thinks the schema should be included in the ProducerOptions/ConsumerOptions

Reply:
@merlimat I am not sure about that. 

a) if we are following Java, it is better to have schema as a separated argument.
b) schema is not an `option`. it is a long-live go instance during the lifecycle of a producer/consumer. but options are only used during construction. after they are passed to cpp client's configuration, the go options are not used anymore. 

so I don't think it is a good idea to add schema to the options.

Reply:
> a) if we are following Java, it is better to have schema as a separated argument.

The reason for that is how Java generics work and the interaction with builder. We need to pass the the type `<T>` when the builder is created. 
That won't be the case in Go.

> b) schema is not an option. it is a long-live go instance during the lifecycle of a producer/consumer. but options are only used during construction. after they are passed to cpp client's configuration, the go options are not used anymore.

That's the same as Authentication (in both Go and Java). Passing these interfaces in as options makes the API less cluttered and easier to add things in a backward compatible way.

Reply:
> That's the same as Authentication (in both Go and Java).

I think Authentication is part of Java. so it is okay to use the same approach as Java.

> The reason for that is how Java generics work and the interaction with builder. We need to pass the the type <T> when the builder is created.

sure. go doesn't have generic. but having a similar interface as Java is much better from maintenance perspective. otherwise it is really confused that Java is having schema out side of producer/consumer configuration, but Go is having schema as part of producer/consumer configuration.

so my take on this would make interfaces and apis as similar as possible between different language clients. however if you really have a strong opinion on putting schema as part of configuration, I am fine.

Reply:
It's not a super-important thing. 

I, of course, completely agree on maintaining the same concepts across languages APIS. That's why I was insisting in having the same concept of schema as in Java/Python. 
When we drill down to the lower level though, there are already differences in the ergonomics of the API between Java and Go, and for a good reasons: to maintain a "native" feeling for the API. 

In general, we've been following the path of simplification for Go API (as simpler APIs are generally preferred there) and, in this specific context, I'd rather avoid duplicating the methods (create with schema and create without schema).

`ProducerOption` is really the equivalent of the builder or the args for Python. For example, in python we're doing: 

```python
producer = client.create_producer(
                    topic='my-topic',
                    schema=AvroSchema(Example) )
```
which would be the equivalent of 
```go
producer, err := client.CreateProducer(pulsar.ProducerOptions{
		Topic: "my-topic",
		Schema: NewAvroSchema(...),
	})
```

So I don't know whether `ProducerOptions` is already poorly named or what's the line between configuration and functionality. 
To me, it seems more natural to have a single structure to "declare" how the producer look like. 

Having said that, I can live with both approaches.

Reply:
I am fine. I will leave @wolfstudy to make the decision.

Reply:
thanks @sijie @merlimat , study a lot.

This is a tough choice, both ways are very good implementation. I am currently in accordance with the way of @sijie , all the tests have passed, so, whether we first temporarily follow this implementation idea.

 @merlimat Do you think it is OK?